blockchain/smart_contracts/smart_contracts.py

```diff
@@ -0,0 +1,268 @@
+                sc_key += pl['transaction_type']
+            else:
+                return False
+        sc_key += ":"
+        if "phase" in criteria:
+            if "phase" in pl and pl['phase']:
+                sc_key += str(pl['phase'])
+            else:
+                return False
+        # insert new sc into database
+        if not self._insert_sc(pl, "ssc", sc_key):
+            return False
+        return self._sc_provisioning_helper(pl, "ssc", sc_key)
+
+    # TODO: implement lsc functionality
+    def provision_lsc(self, transaction):
+        """ provide name and python module to run sc """
+        return True
+
+    def provision_bsc(self, transaction):
+        """
+        provision bsc class smart contract
+        :param transaction: transaction to extract sc data from
+        """
+        status = False
+        pl = transaction['payload']
+        if "phase" in pl and pl['phase']:
+            sc_key = pl['phase']
+            if self._insert_sc(pl, "bsc", sc_key):
+                status = self._sc_provisioning_helper(pl, "bsc", sc_key)
+        return status
+
+    def _sc_provisioning_helper(self, pl, sc_class, sc_key):
+        """
+        insert sc code into appropriate dictionary
+        :param pl: transaction payload to extract sc from
+        :param sc_class: type of sc being dealt with (e.g. tsc, ssc, etc.)
+        :param sc_key: transaction type
+        """
+        try:
+            if 'smart_contract' in pl:
+                sc = pl['smart_contract']
+                if sc[sc_class]:
+                    func = None
+                    # define sc function
+                    exec(sc[sc_class])
+                    # store sc function for this txn type (sc_key)
+                    self.sc_container[sc_class][sc_key] = func
+                else:
+                    logger().warning("No smart contract code provided...")
+                    return False
+            else:
+                return False
+        except Exception as ex:
+            template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+            message = template.format(type(ex).__name__, ex.args)
+            logger().warning(message)
+            return False
+        return True
+
+    def execute_tsc(self, transaction):
+        txn_type = transaction["header"]["transaction_type"]
+        try:
+            self.tsc[txn_type](transaction)
+        except:
+            logger().warning("An error occurred during tsc execution on transaction: %s", transaction['owner'])
+            return False
+        return True
+
+    def execute_ssc(self, min_block_id):
+        """ execute subscription smart contract """
+        for sc_key in self.ssc.keys():
+            (origin_id, txn_type, phase) = sc_key.split(":")
+            vrs = verification_db.get_all(origin_id=origin_id, phase=phase, min_block_id=min_block_id)
+            for v in vrs:
+                txns = transaction_db.get_all(txn_type, v['block_id'])
```

Comment:
You added the origin, but you still didn't address the redundancy issue...

Here's an example scenario:
I subscribe to a server and require 10 p2 verifications per block. During a 5 second window, the server I'm subscribed to processes 100 txns that match the subscription. Using this approach, you'll end up fetching the same set of 100 txns 10 times (10 vrs). If you know that the vrs are all grouped with the same set of transactions (block + origin), you could get away with only fetching 100 instead of 1000 txns by reusing them for multiple smart contract executions.